### PR TITLE
Better sound occlusion

### PR DESCRIPTION
**Pull request by Clyybber**

### CODE REVIEW

File: Robust.Shared/Physics/PhysicsManager.cs [hunk 1]
```diff
@@ -276,6 +276,26 @@ public void RemoveBody(IPhysBody physBody)
         public IEnumerable<RayCastResults> IntersectRay(MapId mapId, CollisionRay ray, float maxLength = 50, IEntity ignoredEnt = null, bool returnOnFirstHit = true)
             => IntersectRayWithPredicate(mapId, ray, maxLength, entity => entity == ignoredEnt, returnOnFirstHit);
 
+        /// <inheritdoc />
+        public float IntersectRayPenetration(MapId mapId, CollisionRay ray, float maxLength, IEntity ignoredEnt = null)
+        {
+            var penetration = 0f;
+
+            var sourceToDest = (List<RayCastResults>) IntersectRay(mapId, ray, maxLength, ignoredEnt, false);
```
Comment (ComicIronic) [inline Robust.Shared/Physics/PhysicsManager.cs]: I don't think you should cast these. If you need list-y features, you should either create a list or change the return interface of `IntersectRay`.

Comment (ComicIronic) [reply Robust.Shared/Physics/PhysicsManager.cs]: Or alternatively use a private function with return type `List` that the public function `IntersectRay` wraps.

Comment (RemieRichards) [reply Robust.Shared/Physics/PhysicsManager.cs]: `IEnumerable<T>` actually *does* have `.Count()` available as an extension method, and `.Count()` seems to be all @Clyybber wants.

(Yes it's a method, `.Count()` not the same as the property `.Count`)

Comment (Clyybber) [reply Robust.Shared/Physics/PhysicsManager.cs]: I also need to index elements. I simply used `.ToArray` now

Comment (RemieRichards) [reply Robust.Shared/Physics/PhysicsManager.cs]: Ah fair enough, missed that.

File: Robust.Shared/Physics/PhysicsManager.cs [hunk 2]
```diff
@@ -276,6 +276,26 @@ public void RemoveBody(IPhysBody physBody)
...
+        public float IntersectRayPenetration(MapId mapId, CollisionRay ray, float maxLength, IEntity ignoredEnt = null)
+        {
+            var penetration = 0f;
+
+            var sourceToDest = IntersectRay(mapId, ray, maxLength, ignoredEnt, false).ToArray();
+            var destToSource = IntersectRay(mapId,
+                new CollisionRay(ray.Position + ray.Direction * maxLength, -ray.Direction, ray.CollisionMask),
+                maxLength, ignoredEnt, false).ToArray();
+
+            if (sourceToDest.Length != destToSource.Length) Logger.Error("This should never occur.");
```
Comment (gradientvera) [inline Robust.Shared/Physics/PhysicsManager.cs]: If that should never occur, why not throw an exception instead?

Comment (Clyybber) [reply Robust.Shared/Physics/PhysicsManager.cs]: Uses `Debug.Assert` now

**Review by gradientvera — APPROVED** [review]

**Pull request merged**